### PR TITLE
Peek reader

### PR DESCRIPTION
**Pull request by Stargateur**
This add a way to not directly consume, useful with crate like bincode and networking when you can't know if all data require is present.

For example, you can feed the peek reader to bincode and validate the number of bytes readed later using `count_peek()` to get the number of bytes used by bincode on success.

### CODE REVIEW

**Review by jeromefroe — COMMENTED** [review]
This is great, thank you @Stargateur! Mind just adding a test for the peak reader to ensure it doesn't advance the read cursor?

**Comment by Stargateur** [comment]
oops, I didn't think about infinite loop. I work on a solution.

**Comment by Stargateur** [comment]
Should work now, thus I'm still not happy with the code, specially the unsafe part I wonder if a safe replace of copy_overlap could be use, I didn't search yet.

I choice to factorize as an associated function instead of a method cause it's avoid any mistake about use internal read_cursor instead of the peek_cursor.

**Comment by Stargateur** [comment]
unresolved question, should `reader_peek()` take self as mutable to allow a `consume()` fn to simply let the user to do `reader_peek.consume()` to advance to real read cursor.

```rust
fn consume(&mut self) {
  self.inner.read_cursor = self.peek_cursor;
}
```

File: src/lib.rs [hunk 1]
```diff
@@ -479,39 +487,103 @@ impl CircBuf {
...
+        )
+    }
+
+    /// Reset the peek cursor to the original read cursor of the inner `CircBuf`
+    pub fn reset(&mut self) {
+        self.peek_cursor = self.inner.read_cursor;
+    }
+
+    /// Return the number of bytes currently peeked. Useful to use for `CircBuf::advance_read`.
+    pub fn readed(&self) -> usize {
```
Comment (jeromefroe) [inline src/lib.rs]: nit: What do you think about naming this just `read`?

**Review by jeromefroe — COMMENTED** [review]
Changes look good to me overall, thanks for taking the time and effort to make this PR! Had just one small comment/question.

File: src/lib.rs (hunk 1 above)
Comment (Stargateur) [reply src/lib.rs]: I don't like much `read`, it's not clear enough. `used()` ? `count()` ? `count_read()`/`read_count()` ? `count_peek()` ? `bytes_peek()`/`bytes_read` ?

Comment (jeromefroe) [reply src/lib.rs]: Good point, maybe `count_peek` then?

**Comment by Stargateur** [comment]
Rename done, fixed some warning and rebase commit, I did a small perf improvement of `is_empty()` and added an `is_empty()` on the peek reader.

**Review by jeromefroe — APPROVED** [review]
Nice!

**Pull request merged**